v23.1/transaction-retry-error-example.md

```diff
@@ -0,0 +1,118 @@
+                raise e
+
+        raise ValueError(f"Transaction did not succeed after {max_retries} retries")
+~~~
+
+`run_transaction` takes a SQL-executing function `op(conn)` and attempts to run the function, retrying on serialization failures (exposed in pscyopg2 as the [`SerializationFailure` exception class](https://www.psycopg.org/docs/errors.html#sqlstate-exception-classes)) with exponential backoff, until reaching a maximum number of tries.
+
+You can add a quick test to this function using the `inject_retry_errors_enabled` session variable.
+
+~~~ python
+def run_transaction(conn, op=None, max_retries=3):
+    """
+    Execute the operation *op(conn)* retrying serialization failure.
+    If no op is specified, the function runs a test, using the
+    inject_retry_errors_enabled session variable to inject errors.
+
+    If the database returns an error asking to retry the transaction, retry it
+    *max_retries* times before giving up (and propagate it).
+    """
+    # leaving this block the transaction will commit or rollback
+    # (if leaving with an exception)
+    with conn:
+        for retry in range(1, max_retries + 1):
+            try:
+                if not op:
+                    with conn.cursor() as cur:
+                        if retry == 1:
+                            cur.execute("SET inject_retry_errors_enabled = 'true'")
+                        if retry == max_retries:
+                            cur.execute("SET inject_retry_errors_enabled = 'false'")
+                        cur.execute("SELECT now()")
+                        logging.debug("status message: %s", cur.statusmessage)
+                else:
+                    op(conn)
+                # If we reach this point, we were able to commit, so we break
+                # from the retry loop.
+                return
+
+            except SerializationFailure as e:
+                # This is a retry error, so we roll back the current
+                # transaction and sleep for a bit before retrying. The
+                # sleep time increases for each failed transaction.
+                logging.debug("got error: %s", e)
+                conn.rollback()
+                sleep_ms = (2 ** retry) * 0.1 * (random.random() + 0.5)
+                logging.debug("Sleeping %s seconds", sleep_ms)
+                time.sleep(sleep_ms)
+
+            except psycopg2.Error as e:
+                logging.debug("got error: %s", e)
+                raise e
+
+        raise ValueError(f"Transaction did not succeed after {max_retries} retries")
+~~~
+
+Calling `run_transaction` without an `op` input sets `inject_retry_errors_enabled` as `true` until the final retry attempt, before which the `inject_retry_errors_enabled` is set back to `false`. For all attempts except the last one, CockroachDB will inject a retryable serialization error for the client to handle. If the client cannot handle the error properly, the retry logic isn't working properly.
```

Comment:
The error handling bit, after "For all..." seems separate and important enough to have its own header, nah?